### PR TITLE
Copy event

### PR DESCRIPTION
**Pull request by smugikity**
add boolean deep. if true copy all related tasks and eventroles
add test
add @Lazy to EventService in TaskService

### CODE REVIEW

**Comment by IP-13** [comment]
|Overall Project|56.38% **`-1.25%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|71.9%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/dc4baa5f102a41d02ca780cc2542690e366de7a7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|88.01% **`-3.55%`**|:green_apple:|
|[EventController.java](https://github.com/itmo-events-app/event-app/blob/dc4baa5f102a41d02ca780cc2542690e366de7a7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FEventController.java)|86.9% **`-13.1%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/dc4baa5f102a41d02ca780cc2542690e366de7a7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|72.53% **`-2.65%`**|:green_apple:|

File: src/main/java/org/itmo/eventapp/main/service/EventService.java [hunk 1]
```diff
@@ -204,4 +206,42 @@ public boolean checkOneEvent(Event first, Event second) {
     public List<EventRole> getUsersHavingRoles(Integer id) {
         return eventRoleService.findAllByEventId(id);
     }
+
+    public Event copyEvent(int id, boolean deep) {
```
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:210]: Добавь еще `//TODO: Add privilege validation`

File: src/main/java/org/itmo/eventapp/main/service/EventService.java [hunk 2]
```diff
@@ -204,4 +206,42 @@ public boolean checkOneEvent(Event first, Event second) {
...
+                .startDate(existingEvent.getStartDate())
+                .endDate(existingEvent.getEndDate())
+                .title(existingEvent.getTitle())
+                .shortDescription(existingEvent.getShortDescription())
+                .fullDescription(existingEvent.getFullDescription())
+                .format(existingEvent.getFormat())
+                .status(existingEvent.getStatus())
+                .registrationStart(existingEvent.getRegistrationStart())
+                .registrationEnd(existingEvent.getRegistrationEnd())
+                .parent(existingEvent)
```
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:223]: Так, нет, тут этого не должно быть. Нужно взять `existingEvent.getParent()`

File: src/main/java/org/itmo/eventapp/main/service/EventService.java (hunk 1 above)
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:210]: `@Transactional`

File: src/main/java/org/itmo/eventapp/main/service/EventService.java [hunk 3]
```diff
@@ -204,4 +206,42 @@ public boolean checkOneEvent(Event first, Event second) {
...
+            for (EventRole eventRole : eventRoles) {
+                EventRole copiedEventRole = EventRole.builder()
+                        .event(savedEvent)
+                        .user(eventRole.getUser())
+                        .role(eventRole.getRole())
+                        .build();
+                eventRoleService.save(copiedEventRole);
+            }
+            List<Integer> taskIds = taskService.findAllByEventId(existingEvent.getId()).stream().map(Task::getId).collect(Collectors.toList());
+            taskService.copyTasks(savedEvent.getId(),taskIds);
```
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:243]: Задачи не копируем

File: src/main/java/org/itmo/eventapp/main/service/EventService.java [hunk 4]
```diff
@@ -204,4 +206,42 @@ public boolean checkOneEvent(Event first, Event second) {
...
+
+        if (deep) {
+            List<EventRole> eventRoles = eventRoleService.findAllByEventId(existingEvent.getId());
+            for (EventRole eventRole : eventRoles) {
+                EventRole copiedEventRole = EventRole.builder()
+                        .event(savedEvent)
+                        .user(eventRole.getUser())
+                        .role(eventRole.getRole())
+                        .build();
+                eventRoleService.save(copiedEventRole);
```
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:240]: Можно ли сделать save 1 раз для всех новых записей? Не ходить в БД в цикле

File: src/main/java/org/itmo/eventapp/main/service/EventService.java [hunk 5]
```diff
@@ -204,4 +206,42 @@ public boolean checkOneEvent(Event first, Event second) {
...
+            }
+            List<Integer> taskIds = taskService.findAllByEventId(existingEvent.getId()).stream().map(Task::getId).collect(Collectors.toList());
+            taskService.copyTasks(savedEvent.getId(),taskIds);
+        }
+        return savedEvent;
```
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:245]: Нету копирования активностей.
Активность – это сущность, у которой parent_id = event.id. Если deep, то нужно сделать
1) eventRepository.findAllByParentId(existingEvent.getId())
2) для каждого event'а из полученного списка создать копию, у которой parentId = newEvent.id

**Review by omgronny — CHANGES_REQUESTED** [review]

File: src/test/java/org/itmo/eventApp/main/service/EventServiceTest.java [hunk 6]
```diff
@@ -159,4 +165,28 @@ void testDeleteEventById() {
...
+            return event;
+        });
+
+        Event copiedEvent = eventService.copyEvent(1, true);
+
+        verify(eventRepository).save(any(Event.class));
+
+        assertEquals(existingEvent.getStartDate(), copiedEvent.getStartDate());
+        assertEquals(existingEvent.getEndDate(), copiedEvent.getEndDate());
+        assertEquals(existingEvent.getTitle(), copiedEvent.getTitle());
```
Comment (omgronny) [inline src/test/java/org/itmo/eventApp/main/service/EventServiceTest.java:190]: А давай еще сделаем проверку что existingEvent.getId() != copiedEvent.getId()

File: src/test/java/org/itmo/eventApp/main/service/EventServiceTest.java [hunk 7]
```diff
@@ -159,4 +165,28 @@ void testDeleteEventById() {
         verify(eventRepository).deleteById(eventId);
         verify(eventRepository, times(1)).deleteById(eventId);
     }
+
+    @Test
+    void testCopyEvent() {
```
Comment (thermvs) [inline src/test/java/org/itmo/eventApp/main/service/EventServiceTest.java:170]: Можно добавить проверки для остальных полей объекта, проверки вызова методов findAllByEventId, save и copyTasks через verify, рассмотреть случай, когда deep равен false, и когда existingEvent не найдено в бд

**Comment by IP-13** [comment]
|Overall Project|58.89% **`-1.3%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|74.58%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/01c63af2de3942a771dc65336905074cb602cb2b/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|90% **`-2.79%`**|:green_apple:|
|[EventController.java](https://github.com/itmo-events-app/event-app/blob/01c63af2de3942a771dc65336905074cb602cb2b/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FEventController.java)|86.9% **`-13.1%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/01c63af2de3942a771dc65336905074cb602cb2b/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|72.53% **`-2.65%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/01c63af2de3942a771dc65336905074cb602cb2b/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|15.58% **`-2.17%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|58.84% **`-1.3%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|73.84%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/35a4a53c846bb494b562392b9d6d0c61456a7267/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|89.92% **`-2.81%`**|:green_apple:|
|[EventController.java](https://github.com/itmo-events-app/event-app/blob/35a4a53c846bb494b562392b9d6d0c61456a7267/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FEventController.java)|86.9% **`-13.1%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/35a4a53c846bb494b562392b9d6d0c61456a7267/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|72.53% **`-2.65%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/35a4a53c846bb494b562392b9d6d0c61456a7267/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|15.58% **`-2.17%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|58.84% **`-1.3%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|73.84%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/d8f382a10bbb91652f6bb317ceecc0d0498d062c/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|89.92% **`-2.81%`**|:green_apple:|
|[EventController.java](https://github.com/itmo-events-app/event-app/blob/d8f382a10bbb91652f6bb317ceecc0d0498d062c/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FEventController.java)|86.9% **`-13.1%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/d8f382a10bbb91652f6bb317ceecc0d0498d062c/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|72.53% **`-2.65%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/d8f382a10bbb91652f6bb317ceecc0d0498d062c/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|15.58% **`-2.17%`**|:x:|

**Comment by IP-13** [comment]
|Overall Project|58.84% **`-1.3%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|73.84%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/0266403ad88875108706399124a79e16a21894d7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|89.92% **`-2.81%`**|:green_apple:|
|[EventController.java](https://github.com/itmo-events-app/event-app/blob/0266403ad88875108706399124a79e16a21894d7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FEventController.java)|86.9% **`-13.1%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/0266403ad88875108706399124a79e16a21894d7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|72.53% **`-2.65%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/0266403ad88875108706399124a79e16a21894d7/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|15.58% **`-2.17%`**|:x:|

File: src/main/java/org/itmo/eventapp/main/service/EventService.java [hunk 8]
```diff
@@ -204,4 +206,59 @@ public boolean checkOneEvent(Event first, Event second) {
...
+    }
+
+    @Transactional
+    public void saveAll(List<Event> events) {
+        eventRepository.saveAll(events);
+    }
+
+    @Transactional
+    public Event copyEventByOne(Event existingEvent, Event parentEvent) {
+        Event copiedEvent = Event.builder()
```
Comment (omgronny) [inline src/main/java/org/itmo/eventapp/main/service/EventService.java:232]: У нас есть EventMapper, надо тут воспользоваться им

**Comment by IP-13** [comment]
|Overall Project|54.17% **`-1.21%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|64.29%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/8f9b4bd5818816f8ada2cfbc92ea30cdd4228aeb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|88.38% **`-2.99%`**|:green_apple:|
|[EventController.java](https://github.com/itmo-events-app/event-app/blob/8f9b4bd5818816f8ada2cfbc92ea30cdd4228aeb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FEventController.java)|86.9% **`-13.1%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/8f9b4bd5818816f8ada2cfbc92ea30cdd4228aeb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|63.35% **`-2.28%`**|:green_apple:|
|[EventRoleService.java](https://github.com/itmo-events-app/event-app/blob/8f9b4bd5818816f8ada2cfbc92ea30cdd4228aeb/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventRoleService.java)|15.58% **`-2.17%`**|:x:|

**Review by omgronny — APPROVED** [review]

**Review by thermvs — APPROVED** [review]

**Pull request merged**